canopen/profiles/p402.py

```diff
@@ -45,59 +201,214 @@ class Node402(RemoteNode):
     """
 
     def __init__(self, node_id, object_dictionary):
-        super(Node402, self).__init__(node_id, object_dictionary)
-        self.powerstate_402 = PowerStateMachine(self)
-        self.powerstate_402.network = self.network
+        super(BaseNode402, self).__init__(node_id, object_dictionary)
+
+        self.is_statusword_configured = False
+
+        #: List of values obtained by the configured TPDOs in a dictionary {object (hex), value}
+        self.tpdo_values = {}
+        #! list of mapped objects configured in the RPDOs in a dictionary {object (hex, pointer (RPDO object) }
+        self.rpdo_pointers = {}
 
     def setup_402_state_machine(self):
-        # setup TPDO1 for this node
-        # TPDO1 will transmit the statusword of the 402 control state machine
-        # first read the current PDO setup and only change TPDO1
-        print(self.nmt.state)
+        """Configured the state machine by searching for the PDO that has the
+        StatusWord mappend.
+        :raise ValueError: If the the node can't finde a Statusword configured
+        in the any of the TPDOs
+        """
+        # the node needs to be in pre-operational mode
         self.nmt.state = 'PRE-OPERATIONAL'
-        self.tpdo[1].read()
-        self.tpdo[1].clear()
-        # Use register as to stay manufacturer agnostic
-        self.tpdo[1].add_variable(0x6041)
-        # add callback to listen to TPDO1 and change 402 state
-        self.tpdo[1].add_callback(self.powerstate_402.on_PDO1_callback)
-        self.tpdo[1].trans_type = 255
-        self.tpdo[1].enabled = True
-        self.tpdo[1].save()
+        self.pdo.read()  # read all the PDOs (TPDOs and RPDOs)
+        #
+        for tpdo in self.tpdo.values():
+            if tpdo.enabled:
+                tpdo.add_callback(self.on_TPDOs_update_callback)
+                for obj in tpdo:
+                    logger.debug('Configured TPDO: {0}'.format(obj.index))
+                    if obj.index not in self.tpdo_values:
+                        self.tpdo_values[obj.index] = 0
+        #
+        for rpdo in self.rpdo.values():
+            for obj in rpdo:
+                logger.debug('Configured RPDO: {0}'.format(obj.index))
+                if obj.index not in self.rpdo_pointers:
+                    self.rpdo_pointers[obj.index] = obj
+
+        # Check if the Controlword is configured
+        if 0x6040 not in self.rpdo_pointers:
+            raise ValueError('Controlword not configured in the PDOs of this node, using SDOs to set Controlword')
```

Comment:
Should this really be an error ? Wouldn't it be better to simply log a warning? 

The `ValueError` implies that the function failed but the setter for the control word checks whether or not the control word is in the `rpdo_pointers` which seems to indicate that this should not be an error.

Reply:
I think you are right, we could change that to an error. Also if the the Controlword (0x6040) is not configured in the PDO it will try to use the SDO's (more slowly) to set the Controlword. I will make a push request to change this.